Reduce header includes in LocalDOMWindow.h

rniwa opened the pull request:
#### 861940cc5a60857216cd3b21583816717e7706db
<pre>
Reduce header includes in LocalDOMWindow.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=292655">https://bugs.webkit.org/show_bug.cgi?id=292655</a>

Reviewed by Brent Fulgham.

Reduced #include in LocalDOMWindow.h to improve the build time.

This PR also merges Base64Utilities into WindowOrWorkerGlobalScope and make LocalDOMWindow
lazily construct PushManager in LocalDOMWindow::pushManager instead of eagerly constructing it
in LocalDOMWindow::LocalDOMWindow.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/push-api/PushSubscriptionOwner.h:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
(WebCore::DialogHandler::dialogCreated):
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/dom/KeyboardEvent.cpp:
(WebCore::viewIsCompositing):
(WebCore::KeyboardEvent::charCode const):
* Source/WebCore/dom/MouseRelatedEvent.cpp:
(WebCore::MouseRelatedEvent::frameViewFromWindowProxy):
* Source/WebCore/history/CachedFrame.h:
* Source/WebCore/history/CachedPage.h:
* Source/WebCore/html/PDFDocument.cpp:
* Source/WebCore/page/Base64Utilities.cpp: Removed.
* Source/WebCore/page/Base64Utilities.h: Removed.
* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::btoa):
(WebCore::DOMWindow::atob):
* Source/WebCore/page/DOMWindowExtension.cpp:
(WebCore::DOMWindowExtension::frame const):
* Source/WebCore/page/DeviceController.cpp:
* Source/WebCore/page/Frame.cpp:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::dispatchAllPendingBeforeUnloadEvents):
(WebCore::LocalDOMWindow::allowPopUp):
(WebCore::ContextDestructionObserver):
(WebCore::LocalDOMWindow::isCurrentlyDisplayedInFrame const):
(WebCore::LocalDOMWindow::collectMatchingElementsInFlatTree):
(WebCore::LocalDOMWindow::matchingElementInFlatTree):
(WebCore::LocalDOMWindow::orientation const):
(WebCore::LocalDOMWindow::processPostMessage):
(WebCore::LocalDOMWindow::focus):
(WebCore::LocalDOMWindow::blur):
(WebCore::LocalDOMWindow::print):
(WebCore::LocalDOMWindow::stop):
(WebCore::LocalDOMWindow::alert):
(WebCore::LocalDOMWindow::confirmForBindings):
(WebCore::LocalDOMWindow::prompt):
(WebCore::LocalDOMWindow::find const):
(WebCore::LocalDOMWindow::outerHeight const):
(WebCore::LocalDOMWindow::outerWidth const):
(WebCore::LocalDOMWindow::innerHeight const):
(WebCore::LocalDOMWindow::innerWidth const):
(WebCore::LocalDOMWindow::screenX const):
(WebCore::LocalDOMWindow::screenY const):
(WebCore::LocalDOMWindow::scrollX const):
(WebCore::LocalDOMWindow::scrollY const):
(WebCore::LocalDOMWindow::name const):
(WebCore::LocalDOMWindow::setName):
(WebCore::LocalDOMWindow::disownOpener):
(WebCore::LocalDOMWindow::getMatchedCSSRules const):
(WebCore::LocalDOMWindow::scrollBy const):
(WebCore::LocalDOMWindow::scrollTo const):
(WebCore::LocalDOMWindow::allowedToChangeWindowGeometry const):
(WebCore::LocalDOMWindow::incrementScrollEventListenersCount):
(WebCore::LocalDOMWindow::decrementScrollEventListenersCount):
(WebCore::LocalDOMWindow::dispatchLoadEvent):
(WebCore::LocalDOMWindow::dispatchEvent):
(WebCore::LocalDOMWindow::finishedLoading):
(WebCore::LocalDOMWindow::setLocation):
(WebCore::LocalDOMWindow::open):
(WebCore::LocalDOMWindow::showModalDialog):
(WebCore::LocalDOMWindow::frame const):
(WebCore::LocalDOMWindow::localFrame const):
(WebCore::LocalDOMWindow::protectedFrame const):
(WebCore::LocalDOMWindow::pushManager):
(WebCore::toScope):
(WebCore::m_pushManager): Deleted.
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/LocalDOMWindowProperty.cpp:
(WebCore::LocalDOMWindowProperty::frame const):
* Source/WebCore/page/Location.cpp:
(WebCore::Location::replace):
(WebCore::Location::setLocation):
* Source/WebCore/page/NavigationHistoryEntry.h:
* Source/WebCore/page/WebKitNamespace.cpp:
* Source/WebCore/page/WindowOrWorkerGlobalScope.cpp:
(WebCore::WindowOrWorkerGlobalScope::btoa):
(WebCore::WindowOrWorkerGlobalScope::atob):
* Source/WebCore/page/WindowOrWorkerGlobalScope.h:
* Source/WebCore/testing/Internals.cpp:
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::remotePostMessage):

Canonical link: <a href="https://commits.webkit.org/294653@main">https://commits.webkit.org/294653@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11f23aa02cc5d51a82b873ad6eab2ab4ec527c90

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102536 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22205 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12519 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107697 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53173 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104575 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22513 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30711 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78000 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34991 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105542 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17440 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92551 "Found 1 new API test failure: TestWebKitAPI.WebKit.PrintFrame (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58336 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17278 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10579 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52530 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87093 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10648 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110072 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29668 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21879 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86980 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30032 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88745 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86577 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31410 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9135 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23918 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16659 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29596 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34908 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29407 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32730 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30966 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2785,15 +2785,20 @@ void LocalDOMWindow::disableSuddenTermination()
-LocalFrame* LocalDOMWindow::frame() const
+Frame* LocalDOMWindow::frame() const
+{
+    return localFrame();
```
Is this just here for backwards compatibility? Maybe it would enforce better coding if we removed it?

rniwa replied:
Yeah, I'll do a follow up.

rniwa replied:
Actually, we can't get rid of this override because there are clients relying on `DOMWindow::frame` which is implemented by this class and `RemoteDOMWindow`.

brentfulgham commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2823,12 +2828,14 @@ CloseWatcherManager& LocalDOMWindow::closeWatcherManager()
 PushManager& LocalDOMWindow::pushManager()
 {
-    return m_pushManager;
+    if (!m_pushManager)
```
Nice! I wonder if this is a load time progression.

brentfulgham reviewed (approved):
LGTM.

rniwa committed:
Reduce header includes in LocalDOMWindow.h
https://bugs.webkit.org/show_bug.cgi?id=292655

Reviewed by Brent Fulgham.

Reduced #include in LocalDOMWindow.h to improve the build time.

This PR also merges Base64Utilities into WindowOrWorkerGlobalScope and make LocalDOMWindow
lazily construct PushManager in LocalDOMWindow::pushManager instead of eagerly constructing it
in LocalDOMWindow::LocalDOMWindow.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/push-api/PushSubscriptionOwner.h:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
(WebCore::DialogHandler::dialogCreated):
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/dom/KeyboardEvent.cpp:
(WebCore::viewIsCompositing):
(WebCore::KeyboardEvent::charCode const):
* Source/WebCore/dom/MouseRelatedEvent.cpp:
(WebCore::MouseRelatedEvent::frameViewFromWindowProxy):
* Source/WebCore/history/CachedFrame.h:
* Source/WebCore/history/CachedPage.h:
* Source/WebCore/html/PDFDocument.cpp:
* Source/WebCore/page/Base64Utilities.cpp: Removed.
* Source/WebCore/page/Base64Utilities.h: Removed.
* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::btoa):
(WebCore::DOMWindow::atob):
* Source/WebCore/page/DOMWindowExtension.cpp:
(WebCore::DOMWindowExtension::frame const):
* Source/WebCore/page/DeviceController.cpp:
* Source/WebCore/page/Frame.cpp:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::dispatchAllPendingBeforeUnloadEvents):
(WebCore::LocalDOMWindow::allowPopUp):
(WebCore::ContextDestructionObserver):
(WebCore::LocalDOMWindow::isCurrentlyDisplayedInFrame const):
(WebCore::LocalDOMWindow::collectMatchingElementsInFlatTree):
(WebCore::LocalDOMWindow::matchingElementInFlatTree):
(WebCore::LocalDOMWindow::orientation const):
(WebCore::LocalDOMWindow::processPostMessage):
(WebCore::LocalDOMWindow::focus):
(WebCore::LocalDOMWindow::blur):
(WebCore::LocalDOMWindow::print):
(WebCore::LocalDOMWindow::stop):
(WebCore::LocalDOMWindow::alert):
(WebCore::LocalDOMWindow::confirmForBindings):
(WebCore::LocalDOMWindow::prompt):
(WebCore::LocalDOMWindow::find const):
(WebCore::LocalDOMWindow::outerHeight const):
(WebCore::LocalDOMWindow::outerWidth const):
(WebCore::LocalDOMWindow::innerHeight const):
(WebCore::LocalDOMWindow::innerWidth const):
(WebCore::LocalDOMWindow::screenX const):
(WebCore::LocalDOMWindow::screenY const):
(WebCore::LocalDOMWindow::scrollX const):
(WebCore::LocalDOMWindow::scrollY const):
(WebCore::LocalDOMWindow::name const):
(WebCore::LocalDOMWindow::setName):
(WebCore::LocalDOMWindow::disownOpener):
(WebCore::LocalDOMWindow::getMatchedCSSRules const):
(WebCore::LocalDOMWindow::scrollBy const):
(WebCore::LocalDOMWindow::scrollTo const):
(WebCore::LocalDOMWindow::allowedToChangeWindowGeometry const):
(WebCore::LocalDOMWindow::incrementScrollEventListenersCount):
(WebCore::LocalDOMWindow::decrementScrollEventListenersCount):
(WebCore::LocalDOMWindow::dispatchLoadEvent):
(WebCore::LocalDOMWindow::dispatchEvent):
(WebCore::LocalDOMWindow::finishedLoading):
(WebCore::LocalDOMWindow::setLocation):
(WebCore::LocalDOMWindow::open):
(WebCore::LocalDOMWindow::showModalDialog):
(WebCore::LocalDOMWindow::frame const):
(WebCore::LocalDOMWindow::localFrame const):
(WebCore::LocalDOMWindow::protectedFrame const):
(WebCore::LocalDOMWindow::pushManager):
(WebCore::toScope):
(WebCore::m_pushManager): Deleted.
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/LocalDOMWindowProperty.cpp:
(WebCore::LocalDOMWindowProperty::frame const):
* Source/WebCore/page/Location.cpp:
(WebCore::Location::replace):
(WebCore::Location::setLocation):
* Source/WebCore/page/NavigationHistoryEntry.h:
* Source/WebCore/page/WebKitNamespace.cpp:
* Source/WebCore/page/WindowOrWorkerGlobalScope.cpp:
(WebCore::WindowOrWorkerGlobalScope::btoa):
(WebCore::WindowOrWorkerGlobalScope::atob):
* Source/WebCore/page/WindowOrWorkerGlobalScope.h:
* Source/WebCore/testing/Internals.cpp:
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::remotePostMessage):

Canonical link: https://commits.webkit.org/294653@main
